File: sample-example.ts
import Libp2p from 'libp2p';
import Kademlia from 'libp2p-kad-dht';
import { Key } from 'libp2p-kad-dht/src/record';

async function main() {
  // Create a libp2p node with the Kademlia DHT module
  const node = await Libp2p.create({
    modules: {
      transport: [],
      streamMuxer: [],
      connEncryption: [],
      dht: Kademlia,
    },
  });

  // Start the node
  await node.start();

  // Insert a (key, value) pair into the DHT
  const key = new Key('my-key');
  const value = Buffer.from('my-value');
  await node.dht.put(key, value);

  // Look up the value for the given key
  const record = await node.dht.get(key);
  const retrievedValue = record.value;
  console.log(`Retrieved value: ${retrievedValue}`);

  // Stop the node
  await node.stop();
}

main();

File: sample-example.rs
use libp2p::kad::record::{self, Key};
use libp2p::kad::{store::MemoryStore, Kademlia};
use std::env;
use std::time::Duration;

#[tokio::main]
async fn main() {
    // Parse command line arguments
    let key = env::args().nth(1).unwrap();
    let value = env::args().nth(2).unwrap();

    // Create a Kademlia DHT with a memory store
    let mut dht = Kademlia::new(env::args().nth(3).unwrap().parse().unwrap(), MemoryStore::new());

    // Insert the (key, value) pair into the DHT
    dht.store(Key::new(key.as_bytes()), value.into(), Duration::from_secs(3600)).await;

    // Look up the value for the given key
    let record = dht.get(Key::new(key.as_bytes())).await.unwrap();
    let value = record::value(&record).unwrap();
    println!("Retrieved value: {}", value);
}
